File: lib/data/repositories/prisoner_repository.dart
import 'package:sqflite/sqflite.dart';
import 'package:uuid/uuid.dart';

import '../../core/constants/app_constants.dart';
import '../../core/constants/enums.dart';
import '../database/database_service.dart';
import '../models/prisoner_model.dart';

// ── Abstract interface ────────────────────────────────────────────────────────

/// All repository methods that return lists accept an optional [station] param.
/// When [station] is non-null only records for that police station are returned.
/// Pass null (admin path) to get records across all stations.
abstract class PrisonerRepositoryBase {
  Future<List<PrisonerModel>> getAll({String? station});
  Future<PrisonerModel?> getById(String id);
  Future<String> insert(PrisonerModel prisoner);
  Future<void> update(PrisonerModel prisoner);
  Future<void> delete(String id);
  Future<List<PrisonerModel>> search(String query, {String? station});
  Future<List<PrisonerModel>> getByStatus(PrisonerStatus status, {String? station});
  Future<List<PrisonerModel>> getByDateFilter(
    DateFilter filter, {
    DateTime? from,
    DateTime? to,
    String? station,
  });
  Future<Map<String, int>> getDashboardStats({String? station});
  Future<Map<String, int>> bulkImport(
    List<PrisonerModel> prisoners, {
    bool updateExisting,
    bool skipDuplicates,
  });

  /// Cross-station search: returns prisoners matching [query] whose station
  /// differs from [excludeStation]. When [excludeStation] is null all records
  /// are returned (admin path). When [query] is empty returns all matching
  /// the station exclusion.
  Future<List<PrisonerModel>> searchCrossStation(
    String query, {
    String? excludeStation,
  });
}

// ── SQLite implementation ─────────────────────────────────────────────────────

class PrisonerRepository implements PrisonerRepositoryBase {
  final _uuid = const Uuid();
  Database get _db => DatabaseService.instance.db;

  // ── Helpers ─────────────────────────────────────────────────────────────────

  /// Builds a WHERE clause fragment for station scoping.
  /// Returns null components when [station] is null (no filtering).
  ({String? clause, List<dynamic>? args}) _stationClause(String? station) {
    if (station == null) return (clause: null, args: null);
    return (clause: 'police_station = ?', args: [station]);
  }

  /// Combines an optional station clause with additional where conditions.
  ({String? where, List<dynamic>? args}) _combineWhere(
    String? station,
    String? extra,
    List<dynamic>? extraArgs,
  ) {
    final sc = _stationClause(station);
    if (sc.clause == null && extra == null) return (where: null, args: null);
    if (sc.clause == null) return (where: extra, args: extraArgs);
    if (extra == null) return (where: sc.clause, args: sc.args);
    return (
      where: '${sc.clause} AND ($extra)',
      args: [...sc.args!, ...?extraArgs],
    );
  }

  // ── CRUD ────────────────────────────────────────────────────────────────────

  @override
  Future<List<PrisonerModel>> getAll({String? station}) async {
    final w = _combineWhere(station, null, null);
    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: w.where,
      whereArgs: w.args,
      orderBy: 'created_at DESC',
    );
    return maps.map(PrisonerModel.fromMap).toList();
  }

  @override
  Future<PrisonerModel?> getById(String id) async {
    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: 'id = ?',
      whereArgs: [id],
      limit: 1,
    );
    return maps.isEmpty ? null : PrisonerModel.fromMap(maps.first);
  }

  @override
  Future<String> insert(PrisonerModel prisoner) async {
    final id = prisoner.id.isEmpty ? _uuid.v4() : prisoner.id;
    final now = DateTime.now();
    final model = prisoner.copyWith(id: id, createdAt: now, updatedAt: now);
    await _db.insert(
      AppConstants.tablePrisoners,
      model.toMap(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
    return id;
  }

  @override
  Future<void> update(PrisonerModel prisoner) async {
    await _db.update(
      AppConstants.tablePrisoners,
      prisoner.copyWith(updatedAt: DateTime.now()).toMap(),
      where: 'id = ?',
      whereArgs: [prisoner.id],
    );
  }

  @override
  Future<void> delete(String id) async {
    await _db.delete(
      AppConstants.tablePrisoners,
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  // ── Queries ──────────────────────────────────────────────────────────────────

  @override
  Future<List<PrisonerModel>> search(String query, {String? station}) async {
    final q = '%$query%';
    const searchExpr =
        'name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ? OR '
        'crime_number LIKE ? OR police_station LIKE ? OR '
        'ipc_sections LIKE ? OR bns_sections LIKE ?';
    final searchArgs = [q, q, q, q, q, q, q];
    final w = _combineWhere(station, searchExpr, searchArgs);
    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: w.where,
      whereArgs: w.args,
      orderBy: 'name ASC',
    );
    return maps.map(PrisonerModel.fromMap).toList();
  }

  @override
  Future<List<PrisonerModel>> getByStatus(
    PrisonerStatus status, {
    String? station,
  }) async {
    final w = _combineWhere(station, 'status = ?', [status.name]);
    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: w.where,
      whereArgs: w.args,
      orderBy: 'name ASC',
    );
    return maps.map(PrisonerModel.fromMap).toList();
  }

  @override
  Future<List<PrisonerModel>> getByDateFilter(
    DateFilter filter, {
    DateTime? from,
    DateTime? to,
    String? station,
  }) async {
    String? dateExpr;
    List<dynamic>? dateArgs;

    final now = DateTime.now();
    switch (filter) {
      case DateFilter.today:
        final today =
            DateTime(now.year, now.month, now.day).toIso8601String();
        dateExpr = "DATE(admission_date) = DATE('$today')";
        break;
      case DateFilter.thisWeek:
        final startOfWeek = now.subtract(Duration(days: now.weekday - 1));
        dateExpr = 'admission_date >= ?';
        dateArgs = [
          DateTime(startOfWeek.year, startOfWeek.month, startOfWeek.day)
              .toIso8601String(),
        ];
        break;
      case DateFilter.thisMonth:
        dateExpr =
            "strftime('%Y-%m', admission_date) = '${now.year}-${now.month.toString().padLeft(2, '0')}'";
        break;
      case DateFilter.custom:
        if (from != null && to != null) {
          dateExpr = 'admission_date >= ? AND admission_date <= ?';
          dateArgs = [from.toIso8601String(), to.toIso8601String()];
        }
        break;
    }

    final w = _combineWhere(station, dateExpr, dateArgs);
    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: w.where,
      whereArgs: w.args,
      orderBy: 'admission_date DESC',
    );
    return maps.map(PrisonerModel.fromMap).toList();
  }

  @override
  Future<Map<String, int>> getDashboardStats({String? station}) async {
    final sc = _stationClause(station);

    Future<int> count(String? extra, [List<dynamic>? extraArgs]) async {
      String where;
      List<dynamic>? wArgs;
      if (sc.clause != null && extra != null) {
        where = ' WHERE ${sc.clause} AND ($extra)';
        wArgs = [...sc.args!, ...?extraArgs];
      } else if (sc.clause != null) {
        where = ' WHERE ${sc.clause}';
        wArgs = sc.args;
      } else if (extra != null) {
        where = ' WHERE $extra';
        wArgs = extraArgs;
      } else {
        where = '';
        wArgs = null;
      }
      return Sqflite.firstIntValue(await _db.rawQuery(
            'SELECT COUNT(*) FROM ${AppConstants.tablePrisoners}$where',
            wArgs,
          )) ??
          0;
    }

    final total = await count(null);
    final undertrial = await count("status = 'undertrial'");
    final convicted = await count("status = 'convicted'");
    final admitted = await count("DATE(admission_date) = DATE('now')");
    final released = await count("status = 'released'");
    final bail = await count("status = 'bail'");
    final transferred = await count("status = 'transferred'");

    return {
      'total': total,
      'undertrial': undertrial,
      'convicted': convicted,
      'admitted': admitted,
      'released': released,
      'bail': bail,
      'transferred': transferred,
    };
  }

  @override
  Future<Map<String, int>> bulkImport(
    List<PrisonerModel> prisoners, {
    bool updateExisting = false,
    bool skipDuplicates = true,
  }) async {
    int inserted = 0, updated = 0, skipped = 0;
    final batch = _db.batch();

    for (final p in prisoners) {
      final existing = await _db.query(
        AppConstants.tablePrisoners,
        where: 'prisoner_id = ?',
        whereArgs: [p.prisonerId],
        limit: 1,
      );

      if (existing.isNotEmpty) {
        if (updateExisting) {
          batch.update(
            AppConstants.tablePrisoners,
            p.copyWith(updatedAt: DateTime.now()).toMap(),
            where: 'prisoner_id = ?',
            whereArgs: [p.prisonerId],
          );
          updated++;
        } else if (skipDuplicates) {
          skipped++;
        }
      } else {
        final id = _uuid.v4();
        final now = DateTime.now();
        batch.insert(
          AppConstants.tablePrisoners,
          p.copyWith(id: id, createdAt: now, updatedAt: now).toMap(),
          conflictAlgorithm: ConflictAlgorithm.ignore,
        );
        inserted++;
      }
    }

    await batch.commit(noResult: true);
    return {'inserted': inserted, 'updated': updated, 'skipped': skipped};
  }

  @override
  Future<List<PrisonerModel>> searchCrossStation(
    String query, {
    String? excludeStation,
  }) async {
    String? where;
    List<dynamic>? whereArgs;

    if (query.isNotEmpty && excludeStation != null) {
      final q = '%$query%';
      where = 'police_station != ? AND '
          '(name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ? OR '
          'crime_number LIKE ? OR police_station LIKE ?)';
      whereArgs = [excludeStation, q, q, q, q, q];
    } else if (query.isNotEmpty) {
      final q = '%$query%';
      where = 'name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ? OR '
          'crime_number LIKE ? OR police_station LIKE ?';
      whereArgs = [q, q, q, q, q];
    } else if (excludeStation != null) {
      where = 'police_station != ?';
      whereArgs = [excludeStation];
    }

    final maps = await _db.query(
      AppConstants.tablePrisoners,
      where: where,
      whereArgs: whereArgs,
      orderBy: 'created_at DESC',
      limit: 200,
    );
    return maps.map(PrisonerModel.fromMap).toList();
  }
}

File: backend/src/repositories/prisoner.repository.js
const { v4: uuidv4 } = require('uuid');
const { getDb } = require('../config/database');
const TABLE = 'prisoners';

class PrisonerRepository {

  getAll(station = null) {
    if (station) {
      return getDb()
        .prepare(`SELECT * FROM ${TABLE} WHERE LOWER(police_station) = LOWER(?) ORDER BY created_at DESC`)
        .all(station)
        .map(rowToModel);
    }
    return getDb()
      .prepare(`SELECT * FROM ${TABLE} ORDER BY created_at DESC`)
      .all()
      .map(rowToModel);
  }

  getById(id) {
    const row = getDb().prepare(`SELECT * FROM ${TABLE} WHERE id = ?`).get(id);
    return row ? rowToModel(row) : null;
  }

  search(query, station = null) {
    const q = `%${query}%`;
    if (station) {
      return getDb().prepare(`
        SELECT * FROM ${TABLE}
        WHERE (name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ?
            OR crime_number LIKE ? OR police_station LIKE ?
            OR ipc_sections LIKE ? OR bns_sections LIKE ?)
          AND LOWER(police_station) = LOWER(?)
        ORDER BY name ASC
      `).all(q, q, q, q, q, q, q, station).map(rowToModel);
    }
    return getDb().prepare(`
      SELECT * FROM ${TABLE}
      WHERE name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ?
          OR crime_number LIKE ? OR police_station LIKE ?
          OR ipc_sections LIKE ? OR bns_sections LIKE ?
      ORDER BY name ASC
    `).all(q, q, q, q, q, q, q).map(rowToModel);
  }

  getByStatus(status, station = null) {
    if (station) {
      return getDb()
        .prepare(`SELECT * FROM ${TABLE} WHERE status = ? AND LOWER(police_station) = LOWER(?) ORDER BY name ASC`)
        .all(status, station)
        .map(rowToModel);
    }
    return getDb()
      .prepare(`SELECT * FROM ${TABLE} WHERE status = ? ORDER BY name ASC`)
      .all(status)
      .map(rowToModel);
  }

  /**
   * Returns all prisoners from stations other than excludeStation.
   * Used by the cross-station read-only screen.
   */
  searchCrossStation(query, excludeStation = null) {
    const hasQuery = query && query.trim().length > 0;
    const q = `%${query}%`;

    if (excludeStation) {
      if (hasQuery) {
        return getDb().prepare(`
          SELECT * FROM ${TABLE}
          WHERE (name LIKE ? OR prisoner_id LIKE ? OR fir_number LIKE ?
              OR crime_number LIKE ? OR police_station LIKE ?
              OR ipc_sections LIKE ? OR bns_sections LIKE ?)
            AND LOWER(police_station) != LOWER(?)
          ORDER BY name ASC
        `).all(q, q, q, q, q, q, q, excludeStation).map(rowToModel);
      }
      return getDb()
        .prepare(`SELECT * FROM ${TABLE} WHERE LOWER(police_station) != LOWER(?) ORDER BY name ASC`)
        .all(excludeStation)
        .map(rowToModel);
    }

    return hasQuery ? this.search(query) : this.getAll();
  }

  getByDateFilter(filter, from, to, station = null) {
    const db = getDb();
    const sc = station ? ` AND LOWER(police_station) = LOWER(?)` : '';
    let stmt, params;

    switch (filter) {
      case 'today':
        stmt   = `SELECT * FROM ${TABLE} WHERE DATE(admission_date) = DATE('now')${sc} ORDER BY admission_date DESC`;
        params = station ? [station] : [];
        break;
      case 'thisWeek':
        stmt   = `SELECT * FROM ${TABLE} WHERE admission_date >= DATE('now', 'weekday 0', '-7 days')${sc} ORDER BY admission_date DESC`;
        params = station ? [station] : [];
        break;
      case 'thisMonth':
        stmt   = `SELECT * FROM ${TABLE} WHERE strftime('%Y-%m', admission_date) = strftime('%Y-%m', 'now')${sc} ORDER BY admission_date DESC`;
        params = station ? [station] : [];
        break;
      case 'custom':
        stmt   = `SELECT * FROM ${TABLE} WHERE admission_date >= ? AND admission_date <= ?${sc} ORDER BY admission_date DESC`;
        params = station ? [from, to, station] : [from, to];
        break;
      default:
        stmt   = station
          ? `SELECT * FROM ${TABLE} WHERE LOWER(police_station) = LOWER(?) ORDER BY admission_date DESC`
          : `SELECT * FROM ${TABLE} ORDER BY admission_date DESC`;
        params = station ? [station] : [];
    }

    return db.prepare(stmt).all(...params).map(rowToModel);
  }

  getDashboardStats(station = null) {
    const db = getDb();
    const sc = station ? ' AND LOWER(police_station) = LOWER(?)' : '';
    const sp = station ? [station] : [];

    const count = (cond) =>
      db.prepare(`SELECT COUNT(*) as n FROM ${TABLE} WHERE ${cond}${sc}`).get(...sp).n;

    return {
      total:       count('1=1'),
      undertrial:  count("status = 'undertrial'"),
      convicted:   count("status = 'convicted'"),
      admitted:    count("DATE(admission_date) = DATE('now')"),
      released:    count("status = 'released'"),
      bail:        count("status = 'bail'"),
      transferred: count("status = 'transferred'"),
    };
  }

  insert(model) {
    const id  = model.id || uuidv4();
    const now = new Date().toISOString();
    const row = modelToRow({ ...model, id, createdAt: now, updatedAt: now });
    getDb().prepare(`
      INSERT INTO ${TABLE}
        (id, prisoner_id, name, age, gender, fir_number, crime_number,
         police_station, prison_name, admission_date, status, ipc_sections,
         bns_sections, release_date, release_reason, remarks, created_at,
         updated_at, created_by)
      VALUES
        (@id, @prisoner_id, @name, @age, @gender, @fir_number, @crime_number,
         @police_station, @prison_name, @admission_date, @status, @ipc_sections,
         @bns_sections, @release_date, @release_reason, @remarks, @created_at,
         @updated_at, @created_by)
    `).run(row);
    return id;
  }

  update(model) {
    const row = modelToRow({ ...model, updatedAt: new Date().toISOString() });
    getDb().prepare(`
      UPDATE ${TABLE} SET
        prisoner_id = @prisoner_id, name = @name, age = @age,
        gender = @gender, fir_number = @fir_number, crime_number = @crime_number,
        police_station = @police_station, prison_name = @prison_name,
        admission_date = @admission_date, status = @status,
        ipc_sections = @ipc_sections, bns_sections = @bns_sections,
        release_date = @release_date, release_reason = @release_reason,
        remarks = @remarks, updated_at = @updated_at
      WHERE id = @id
    `).run(row);
  }

  delete(id) {
    getDb().prepare(`DELETE FROM ${TABLE} WHERE id = ?`).run(id);
  }

  bulkImport(models, { updateExisting = false, skipDuplicates = true } = {}) {
    const db = getDb();
    let inserted = 0, updated = 0, skipped = 0;

    const insertOne = db.prepare(`
      INSERT OR IGNORE INTO ${TABLE}
        (id, prisoner_id, name, age, gender, fir_number, crime_number,
         police_station, prison_name, admission_date, status, ipc_sections,
         bns_sections, release_date, release_reason, remarks,
         created_at, updated_at, created_by)
      VALUES
        (@id, @prisoner_id, @name, @age, @gender, @fir_number, @crime_number,
         @police_station, @prison_name, @admission_date, @status, @ipc_sections,
         @bns_sections, @release_date, @release_reason, @remarks,
         @created_at, @updated_at, @created_by)
    `);

    const updateOne = db.prepare(`
      UPDATE ${TABLE} SET
        name = @name, age = @age, gender = @gender, fir_number = @fir_number,
        crime_number = @crime_number, police_station = @police_station,
        prison_name = @prison_name, admission_date = @admission_date,
        status = @status, ipc_sections = @ipc_sections,
        bns_sections = @bns_sections, release_date = @release_date,
        release_reason = @release_reason, remarks = @remarks,
        updated_at = @updated_at
      WHERE prisoner_id = @prisoner_id
    `);

    const findByPrisonerId = db.prepare(`SELECT id FROM ${TABLE} WHERE prisoner_id = ?`);

    const run = db.transaction((rows) => {
      for (const model of rows) {
        const now      = new Date().toISOString();
        const existing = findByPrisonerId.get(model.prisonerId);

        if (existing) {
          if (updateExisting) {
            updateOne.run(modelToRow({ ...model, updatedAt: now }));
            updated++;
          } else if (skipDuplicates) {
            skipped++;
          }
        } else {
          insertOne.run(modelToRow({ ...model, id: uuidv4(), createdAt: now, updatedAt: now }));
          inserted++;
        }
      }
    });

    run(models);
    return { inserted, updated, skipped };
  }
}

// ── Row ↔ Model mappers ───────────────────────────────────────────────────────

function rowToModel(row) {
  return {
    id:            row.id,
    prisonerId:    row.prisoner_id,
    name:          row.name,
    age:           row.age,
    gender:        row.gender,
    firNumber:     row.fir_number,
    crimeNumber:   row.crime_number,
    policeStation: row.police_station,
    prisonName:    row.prison_name,
    admissionDate: row.admission_date,
    status:        row.status,
    ipcSections:   row.ipc_sections ? row.ipc_sections.split(',').filter(Boolean) : [],
    bnsSections:   row.bns_sections ? row.bns_sections.split(',').filter(Boolean) : [],
    releaseDate:   row.release_date   || null,
    releaseReason: row.release_reason || null,
    remarks:       row.remarks        || null,
    createdAt:     row.created_at,
    updatedAt:     row.updated_at,
    createdBy:     row.created_by     || null,
  };
}

function modelToRow(m) {
  return {
    id:             m.id            || null,
    prisoner_id:    m.prisonerId    || '',
    name:           m.name          || '',
    age:            m.age           || 0,
    gender:         m.gender        || 'male',
    fir_number:     m.firNumber     || '',
    crime_number:   m.crimeNumber   || '',
    police_station: m.policeStation || '',
    prison_name:    m.prisonName    || '',
    admission_date: m.admissionDate || new Date().toISOString(),
    status:         m.status        || 'undertrial',
    ipc_sections:   Array.isArray(m.ipcSections) ? m.ipcSections.join(',') : (m.ipcSections || ''),
    bns_sections:   Array.isArray(m.bnsSections) ? m.bnsSections.join(',') : (m.bnsSections || ''),
    release_date:   m.releaseDate   || null,
    release_reason: m.releaseReason || null,
    remarks:        m.remarks       || null,
    created_at:     m.createdAt     || new Date().toISOString(),
    updated_at:     m.updatedAt     || new Date().toISOString(),
    created_by:     m.createdBy     || null,
  };
}

module.exports = new PrisonerRepository();
